### ISSUE TITLE
we should return role details when revoke and grant users

### ISSUE THREAD

**Issue by yichengq** [coreos/etcd#3709]
It is not consistent with what we do when we get one user or list all users:

```
github.com/coreos/etcd/client.(*httpAuthUserAPI).GetUser
>>> {"user":"userA"}

github.com/coreos/etcd/client.(*httpAuthUserAPI).GrantUser
>>> {"user":"userA","roles":["roleA","roleB"]}

github.com/coreos/etcd/client.(*httpAuthUserAPI).GetUser
>>> {"user":"userA","roles":[{"role":"roleA","permissions":{"kv":{"read":["/roleA/path1/*","/roleA/path2/*"],"write":[]}}},{"role":"roleB","permissions":{"kv":{"read":null,"write":null}}}]}

github.com/coreos/etcd/client.(*httpAuthUserAPI).RevokeUser
>>> {"user":"userA","roles":["roleA"]}
```

found from #3702 
follow #3278 

@raoofm @MSamman @ark76r

**Comment by ark76r** [comment]
@raoofm @MSamman @yichengq 

My proposal about this issue

Some theory at the beginning

Model of Users, Roles, Privileges has existed for decades and in entity-relationship diagram it would look like this:

    [User]<--->[Role]<---[Privilege]

User has primary key "userName", Role has Primary Key "roleName".
userName and roleName has of course unique constraints.

Between User and Role is relation many-to-many (User can have many Roles, Role can be granted to many Users").
There can be Roles existing without any User (not granted to anybody) and Users without Roles (Users without any privileges).

Important question, what to do if User o Role are deleted.
In case of the User it is rather simple: User is deleted but we do not touch Roles which was granted to this Users (so maybe Role become non-used Role).
In case of the Role deletion (if this Role is granted to at least one User), there are two possible scenarios:
1. Server will reject the request (you can't remove Role granted to some Users) or,
2. Server will remove Role and remove all relationship to this Role from Users to whom the Role was granted.

For sake of the usability (the admin don't want to perform revoke on hundreds users), the second scenario looks beter.
Of course the tool will should probably ask in such case administrator if he really want to remove the Rule granted to some Users (eg. "Removig this rule will revoke it from 125 users, are you sure?")

Because of the constraint between User and Role it should be not possible to grant the User the non-existent Role.

Having such model in my opinion is clear and consistent. Users have only existent and valid Roles. Removing Role revokes it from bunch of the Users. 

What about Privilege. It doesn't have to primary key and exist only in the context of the Rule. In UML terms Rule is composed of Privileges (Privilege is the part of the Rule composition).
Privilege doesn't exist without Rule.
)
Rules are identified only by their name. So two Roles with two different names are two different roles (and you cannot have two roles with the same name, because roleName is unique).

If Rule names are different (they have to be unique), these are different and it doesn't matter what Privileges they contain.

We can in many cases just ignore the Privileges inside the Role. Role is kind of the abstract defining rights of the user (it means, that usually it is good idea to give the roles significant names).

API issues

JSON is contract-less format of data what is very handy in languages like JavaScript but in languages with static/strong type validation (like Java or Go) it can be problematic (what we just have the pleasure to see).

It mean, we have to define types for entities User and Role and use only those entites.

But we get some problem, sometimes it is good to get document-like full data, eg:

User:
  - userName
  - roles:
    - roleA: [privilege, ..., ..., ..]

But sometimes it is better to operate only on roleName level (like in functions Grant and Revoke).
More over, sometimes, returning the content of the Role entity (in place of its name) gives the redundancy and typical problems for it.
What about such case:

ListOfUsers:
 - User:
   - userName: userA
   - roles:
     - roleA: [p1, p2, p3]
     - roleB: [p2, p5]
  - User:
    - userName: userB
    - roles:
      - roleA: [p1,p2, p3]
      - roleB: [p3, p5]

The problem is, that roleB is redefined. Shouldn't be, but if you have the role definition repeated for 100 users, it's is kind of problem if you should check if the definitions are exactly the same (in the area of the security, anyone says, that you should ;) )

Summing up, we have some well defined entities, relations among them, primary keys and strong need to design API which will be friendly for languages with strong type control (or however I should call languages like C++, Java, Go).

My proposal is:

Lets define structures:

User:
 - userName (string)
 - password (string - optional)
 - roles ([]string - optional)
 - grant ([]string - optional)
 - revoke ([]string - optional)

Role:
 - roleName (string)
 .. (resto of the role definition, the present seems to be OK)

RoleList:
  - roles ([]Role)

UserList:
  - users ([]User)
  - usedRoles ([]Role - optional)

In such case almost everywhere API operates on Role and User definition, where in use is only roleName. Definition of the role (Role) is used if someone query for it (GetRole) or in special case, when someone asks for user list and wants to get full information about privileges. In such case he/she gets UserList with optional field userRoles, which may be filled with user definitions.

client API is non-problematic, because we have only finite number of immutable structures.
Server returns always one of the known types.

If client soft is going to process privileges, it can download full set of role definitions and join the entities using roleName as a key.

Transfer of UserList is optimal: each role definition is transmitted only once. client API can have switch to inform the server is it is interested in role definitions or not (if not, usedRoles is empty).

This API gives also one additional advantage: userRoles in UserList enumerate all roles granted to any user. RoleList contains all roles (granted and not granted). Subtraction of these two sets gives set of unused roles (that means maybe roles which should be removed).

In general it looks like in most cases client doesn't need role definition included in User info (event today, the code of the etcdctl looks like it rather doesn't use it, than does). So there is no need to transfer role definitions to the client if we don't know if client is going to use it.

In addition my idea of the UserList can be easily implemented on the server: it passes through user list and builds set of used roles. When finished with users, it goes through collected roleNames and build userRoles list. It could be done even if the server had to stream data to the client without build whole message in memory.

A word about optimizing the number of the round trips to the http server.

Today, in etcdctl, operations 'user grant' and 'user revoke' perform two requests to the server (GetUser and UserGrant/UserRevoke) even if they don't need the role definitions at all. 
So I think it is now more important to design usable and clean API than try to do the best optimized auth/user/role API. In my opinion if someone is going to run 15K requests for user definitions per second, he/she certainly is doing something really wrong.

Disadvantages of such approach: it's more RDBMS-like than document/nosql schema.
But in general tree/document/nosql databases are not too good in modelling many-to-many relationships.

**Comment by ark76r** [comment]
BTW, this is the definition of the User from file etcd/etcdserver/auth/auth.go:

    type User struct {
        User     string   `json:"user"`
        Password string   `json:"password,omitempty"`
        Roles    []string `json:"roles"`
        Grant    []string `json:"grant,omitempty"`
        Revoke   []string `json:"revoke,omitempty"`
    }

I like it. But I don't know where and when, during the Marshall operation, Roles []string is transformed into Roles []Role.

**Comment by xiang90** [comment]
I agree with @ark76r in general. So we will keep the API as it is today.

@ark76r We are designing auth for v3. I will keep you updated and will appreciate your opinions.

/cc @mitake Please /cc @ark76r for related issues. Thanks!

**Comment by mitake** [comment]
@xiang90 ok, I'll cc @ark76r in PRs and issues related to v3 auth.

This is the design doc of the feature: https://docs.google.com/document/d/1Zs5lc4H8kHr4sgoniKlL6YRsLKFNy8m-1pyGhkCmmO4/edit
@ark76r if you have opinions, please comment or update it.

**Comment by ark76r** [comment]
@mitake OK. Give me a day.

**Comment by mitake** [comment]
@ark76r a way of implementing role deletion seems to be interesting:
> Server will remove Role and remove all relationship to this Role from Users to whom the Role was granted.

How about marking a deleted role with a tombstone and cleaning the deleted roles in deferred manner (invoked by user command or periodically)? It will be fast and load (disk I/O, locking of the auth info) can be controlled. In addition, the auth information wouldn't be so large so copy GC can be used for this purpose.

**Comment by ark76r** [comment]
@mitake It was the one of the options (another is to prohibit deletion of the roles granted to the user).
And yes, of course internally deleting is usually performed with the help of some marker describing 'deleted' state. Deletion itself is usually deferred.

Some of my ideas below (only for discussion)

Some obvious (or not) facts:

1. Authorization must be really quick because it has negative impact for all operations (especially accessing elements of the path) 
2. admin operations (eg changing permissions) may may take a relatively long time (authentication also can be moderate slow).
3. There must be defined some consistent model of relations between objects user-role-permission. It also includes answers for questions like 'does user aggregates roles or user consist (among other) of permission set?'
4. The interesting question is: what is there are two users logged in as a administrator and they perform some modifications in parallel? 

Ad 1: during authentication, server should build some kind of effective structure of users permissions (basing on information contained in roles granted to the user). All authorization could be done on this structure. Modifications on the user/role definitions are done on separate structures but when they are finished, server locks users session and recreates its permission set.

Ad 2. It is acceptable (in my opinion!) to force the admin to wait for synchronization of the users/roles definition among the cluster. But he/she don't have to wait for recreating sessions' permission set described above - we only have to be sure that the user/role definition exists on all nodes.

Ad 3. In case of authenticated session the model is user <- (one to many) -> permission and is immutable: exists or is deleted and recreated as a whole. 
User/Role Definitions can have any model - the best is something intuitive and convenient ;) It will have nothing to do with the performance of the authorization process.

Ad 4. We can do some locks on the User/Role Definitions and prevent other session to modify the same data in parallel. It will not impact the performance of the normal read/write data operations.

And another thing: we have session. Maybe the API should be state-ful. That means all write operations could be stored on the server and deffered to the execution of operation named Commit or Apply (or smoething like this).
It is easier to manage set of chagnes at once than many small changes.

Some extra comments to the API described in 'Authentication and Access Control in etcd v3 protocol' (just my _private_ opinions):
Function GetUser should not return all users in any case and RoleGet should not return all roles in any case (why Get-User but Role-Get?).
How many entities may be put into one message?

In my opinion it would be more useful to have GetUser/GetRole returning exactly one item (complete record) and Get(Users|Roles)List to return list of names (strings) matching the pattern given as a parameter. It could also return eg all users having granted some role. (eg to know which users will be affected after changes in role definition).

Questions:
1 - result of the GetUser contains list of roles' names o roles' complete records?
2 - there will be only one admin (root)? 
3 - what ACL model is planned (are you going to give to the admin optional permission to create another admin account? 
4 - who can change permissions: only admin, admin + owner of the object?

For now it looks I know far to less about the plans for v3 ;)

**Comment by mitake** [comment]
@ark76r sorry for my late reply. I need to comment on many opinions and questions from you :) So I'd like to comment on the locking overhead you concern at first.

Overhead of locking auth metadata isn't required. Because every follower can do auth based on its replica of the metadata. Of course it would introduce inconsistent state of the auth. However, actual checking of each request is done by a leader node (based on a token attached to the requests), it must always have the latest auth metadata.

Therefore, auth info update and checking is processed during ordinal raft request sequence. So we don't need locking mechanisms and its overhead.

**Comment by mitake** [comment]
> 1 - result of the GetUser contains list of roles' names o roles' complete records?

I think your suggestion seems good. Fortunately, the APIs aren't implemented yet. I'll separate a functionality of getting a single complete user information from getting entire user list.

> 2 - there will be only one admin (root)? 

AFAIK, v2 auth provides only one admin (root) account. Providing multiple admin accounts useful? I'm not sure.

> 3 - what ACL model is planned (are you going to give to the admin optional permission to create another admin account? 

Ditto.

> 4 - who can change permissions: only admin, admin + owner of the object?

Ditto. And multiple roles can have a permission of single path in our model. Introducing admin permission would result complexity. But zookeeper seems to provide this functionality?
